### ISSUE TITLE
Keeping a reference to a dictionaryIndex leaks objects

### ISSUE THREAD

**Issue by TeunVR** [ReactiveKit/ReactiveKit#22]
Storing the indexForKey in the deletes-array of the ObservableCollectionEvent causes the related value in the dictionary to not be released.
This seems to be a bug in Swift, i filed a bugreport for it: 23952263

```swift
  public mutating func removeValueForKey(key: Index.Key) -> Index.Value? {
    if let index = collection.indexForKey(key) {
      var new = collection
      let oldValue = new.removeValueForKey(key)
      next(ObservableCollectionEvent(collection: new, inserts: [], deletes: [index], updates: []))
      return oldValue
    } else {
      return nil
    }
  }
```

Test-code:
```swift
class MyObject {
    var number : Int
    
    init(nr: Int){
        self.number = nr
    }
    
    deinit{
        print("deinit MyObject \(number)")
    }
}

class Model {
    static let instance = Model()

    var dict = ObservableCollection([1: MyObject(nr:1)])
    
    
    init(){
        Queue.main.after(5.0){
            self.dict[1] = nil
            print("MyObject-1 is not yet released")
            Queue.main.after(5.0){
                print("When adding another object (creating a new CollectionEvent), releases MyObject-1")
                self.dict[2] = MyObject(nr:2)
                Queue.main.after(5.0){
                }
            }
        }
    }
}
```

**Comment by TeunVR** [comment]
[DictionaryIndexLeaksObject.playground.zip](https://github.com/ReactiveKit/ReactiveKit/files/66417/DictionaryIndexLeaksObject.playground.zip)

**Comment by srdanrasic** [comment]
That's an interesting finding. I wander if is has something to do with index being of type  `DictionaryIndex<Key, Value>`. Looks like index is just a key-value pair, although we cannot access neither key nor value.

I'm thinking if it would be possible to fix this on our side. Currently we are storing last event, but that might not be necessary. It should be enough to store just the collection. I'll see if there are any unwanted implications to this.

**Comment by TeunVR** [comment]
It might have something to do with internal Swift optimisations for the copy-on-write mechanism, since Dictionary are value types now. Or some internal caching/reuse of the indexes. 
Anyway, it is weird :)

**Comment by srdanrasic** [comment]
I think we can solve this.

Currently each new observer we get last event that happened. I don't think this is necessary. New observer doesn't really care about last event - only about current state of the collection and future events. Thus, new observers should get just initial event (collection with empty change set array).

If we do that, there is no need to store last event, only last (state of) collection.

**Comment by srdanrasic** [comment]
Just released v1.0.10 with the fix.

**Comment by TeunVR** [comment]
Yep, it's fixed. Thanks!
